app/uk/gov/hmrc/gform/models/FieldValueRaw.scala

```diff
@@ -136,11 +136,19 @@ case class FieldValueRaw(
     o = offset.getOrElse(Offset(0))
   } yield Date(f, o, v)
 
-  private lazy val groupOpt: Opt[Group] = fields.fold(groupOptErr)(optGroup)
-  private lazy val groupOptErr: Opt[Group] = InvalidState(s"""Require 'fields' element in Group""").asLeft
-  private def optGroup(rawFields: List[FieldValueRaw]): Opt[Group] = rawFields.map(_.getFieldValue()).partition(_.isRight) match {
-    case (ueorfvs, Nil) => Group(ueorfvs.map(_.right.get)).asRight
-    case (_, ueorfvs) => ueorfvs.map(_.left.get).head.asLeft
+  private lazy val groupOpt: Either[UnexpectedState, Group] = fields.fold(noRawFields)(createGroupFromRawFieldsOrReturnUnexpectedState)
+  private lazy val noRawFields: Either[UnexpectedState, Group] = InvalidState(s"""Require 'fields' element in Group""").asLeft
+  private def createGroupFromRawFieldsOrReturnUnexpectedState(rawFields: List[FieldValueRaw]): Either[UnexpectedState, Group] = {
```

Comment:
I guess the two sides of the discussion is:
(a) fields.fold(groupOptErr)(optGroup) - requires you to look at the types and the signatures elsewhere to understand what's going on.
(b) fields.fold(noRawFields)(createGroupFromRawFieldsOrReturnUnexpectedState) - requires less of that. Although using the shorter fields.fold(noRawFields)(groupFromRawFieldsOrUnexpectedState) would achieve similar.

Plus, one can observe that once one is very familiar with the types and signatures and the code base in general, the requirement goes away.